File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head == NULL) return head;
        int size = 0;
        ListNode *curr = head;
        while(curr != NULL) {
            curr = curr->next;
            size++;
        }
        if(size < n) return head;

        if(size == n) {
            // remove from head;
            head = head->next;
            return head;
        }
        ListNode *s1 = head, *s2 = head;
        for(int i=0; i<n; i++) {
            s2 = s2->next;
        }

        while(s2->next != NULL) {
            s1 = s1->next;
            s2 = s2->next;
        }

        s1->next = s1->next->next;
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    if(head == null) return head;
    if(head.next == null) return null;
    let first = head, second = head;
    console.log("n: ", n)
    console.log("MOVING SECOND ONLY")
    for(let i=0; i<n; i++) {
        console.log("Second: ", second.val);
        if(second == null) {
            // n nodes aren't present
            return head;
        }
        second = second.next;
    }

    // If `second` is null, we need to remove the head
    if (!second) return head.next;

    console.log("\n\nMOVING BOTH")
    while(second.next != null) {
        console.log("First: ", first.val);
        console.log("Second: ", second.val);
        first = first!.next;
        second = second.next;
    }

    if(first) {
        let skippedNode = first.next; // Can delete this ISOLATED node after in languages like C++

        if(skippedNode) {
            first.next = skippedNode.next;
        }
    }


    return head;
};